File: cppOpen/histoEquali.cpp
#include "opencv2/freetype.hpp"
#include "opencv2/opencv.hpp"
#include <iostream>

using namespace cv;
using namespace std;

String folder = "/home/test/Desktop/opencv/opencv/cppOpen/build/data/";

void drawHistogram(const Mat& src, Mat& histImage) {
    // 히스토그램 설정
    const int histSize = 256; // 히스토그램 크기
    float range[] = { 0, 256 }; // 범위
    const float* histRange = { range };
    
    // 히스토그램 계산
    Mat histogram;
    calcHist(&src, 1, 0, Mat(), histogram, 1, &histSize, &histRange);

    // 히스토그램 이미지 초기화
    histImage = Mat::zeros(300, 512, CV_8UC3);

    // 히스토그램 그리기
    for (int i = 1; i < histSize; i++) {
        line(histImage, Point((i - 1) * 2, 300 - cvRound(histogram.at<float>(i - 1))),
             Point(i * 2, 300 - cvRound(histogram.at<float>(i))),
             Scalar(255, 0, 0), 2, 8, 0);
    }
}

int main() {
    // 이미지 읽기
    Mat src = imread(folder + "img2.bmp", IMREAD_GRAYSCALE);
    if (src.empty()) {
        cout << "이미지를 열 수 없거나 찾을 수 없습니다!" << endl;
        return -1;
    }

    Mat dst;
    // 히스토그램 평활화
    equalizeHist(src, dst);

    // 원본 이미지와 평활화된 이미지 표시
    namedWindow("Original Image", WINDOW_AUTOSIZE);
    imshow("Original Image", src);

    namedWindow("Equalized Image", WINDOW_AUTOSIZE);
    imshow("Equalized Image", dst);

    // 히스토그램 이미지 생성
    Mat histImageSrc, histImageDst;
    drawHistogram(src, histImageSrc);
    drawHistogram(dst, histImageDst);

    // 히스토그램 표시
    namedWindow("Histogram - Original", WINDOW_AUTOSIZE);
    imshow("Histogram - Original", histImageSrc);

    namedWindow("Histogram - Equalized", WINDOW_AUTOSIZE);
    imshow("Histogram - Equalized", histImageDst);

    waitKey(0); // 키 입력 대기
    return 0;
}

File: python_open/histoEquali.py
import cv2
import numpy as np

folder = "/home/test/Desktop/opencv/opencv/cppOpen/build/data/"

def draw_histogram(src):
    # 히스토그램 계산
    hist_size = 256
    hist_range = [0, 256]
    histogram = cv2.calcHist([src], [0], None, [hist_size], hist_range)

    # 히스토그램 이미지 초기화
    hist_image = np.zeros((300, 512, 3), dtype=np.uint8)

    # 히스토그램 그리기
    for i in range(1, hist_size):
        cv2.line(hist_image, (2 * (i - 1), 300 - int(histogram[i - 1])),
                 (2 * i, 300 - int(histogram[i])),
                 (255, 0, 0), 2, 8, 0)
    return hist_image

# 이미지 읽기
src = cv2.imread(folder + "img2.bmp", cv2.IMREAD_GRAYSCALE)
if src is None:
    print("Error: Could not load image!")
    exit(-1)

# 히스토그램 평활화
dst = cv2.equalizeHist(src)

# 원본 이미지와 평활화된 이미지 표시
cv2.namedWindow("Original Image", cv2.WINDOW_AUTOSIZE)
cv2.imshow("Original Image", src)

cv2.namedWindow("Equalized Image", cv2.WINDOW_AUTOSIZE)
cv2.imshow("Equalized Image", dst)

# 히스토그램 이미지 생성
hist_image_src = draw_histogram(src)
hist_image_dst = draw_histogram(dst)

# 히스토그램 표시
cv2.namedWindow("Histogram - Original", cv2.WINDOW_AUTOSIZE)
cv2.imshow("Histogram - Original", hist_image_src)

cv2.namedWindow("Histogram - Equalized", cv2.WINDOW_AUTOSIZE)
cv2.imshow("Histogram - Equalized", hist_image_dst)

cv2.waitKey(0)  # 키 입력 대기
cv2.destroyAllWindows()
